### PR TITLE
Add support for conjugate gradient solver through callable matrix-vector products

### PR DESCRIPTION
**Pull request by KaelanDt**
Adapted from https://jax.readthedocs.io/en/latest/_autosummary/jax.scipy.sparse.linalg.cg.html

### CODE REVIEW

File: posteriors/utils.py [hunk 1]
```diff
@@ -228,6 +230,137 @@ def fisher(*args, **kwargs):
...
+    M: Callable = _identity,
+) -> Tuple[TensorTree, Any]:
+    """Use Conjugate Gradient iteration to solve ``Ax = b``.
+    ``A`` is suplied as a function instead of a sparse
+    matrix or ``LinearOperator``.
+
+    Derivatives of ``cg`` are implemented via implicit differentiation with
+    another ``cg`` solve, rather than by differentiating *through* the solver.
+    They will be accurate only if both solves converge.
+    Adapted from https://jax.readthedocs.io/en/latest/_autosummary/jax.scipy.sparse.linalg.cg.html.
```
Comment (SamDuffield) [inline posteriors/utils.py:287]:
```suggestion
    Adapted from https://jax.readthedocs.io/en/latest/_autosummary/jax.scipy.sparse.linalg.cg.html
```

File: posteriors/utils.py [hunk 2]
```diff
@@ -228,6 +230,137 @@ def fisher(*args, **kwargs):
...
+
+
+def cg(
+    A: Callable,
+    b: TensorTree,
+    x0: TensorTree = None,
+    *,
+    maxiter: int = None,
+    damping: float = 0,
+    tol: Tensor = torch.Tensor([1e-5]),
```
Comment (SamDuffield) [inline posteriors/utils.py:276]: why not just `1e-5`?

File: posteriors/utils.py [hunk 3]
```diff
@@ -228,6 +230,137 @@ def fisher(*args, **kwargs):
...
+_identity = partial(lambda x: x)
+
+
+def cg(
+    A: Callable,
+    b: TensorTree,
+    x0: TensorTree = None,
+    *,
+    maxiter: int = None,
+    damping: float = 0,
```
Comment (SamDuffield) [inline posteriors/utils.py:275]:
```suggestion
    damping: float = 0.0,
```

File: posteriors/utils.py [hunk 4]
```diff
@@ -228,6 +230,137 @@ def fisher(*args, **kwargs):
...
+        info : Placeholder for convergence information. In the future, JAX will report
+            the number of iterations when convergence is not achieved, like SciPy.
+
+    """
+    if x0 is None:
+        x0 = tree_map(torch.zeros_like, b)
+
+    if maxiter is None:
+        size = sum(bi.numel() for bi in tree_leaves(b))
+        maxiter = 10 * size  # copied from scipy
```
Comment (SamDuffield) [inline posteriors/utils.py:317]:
```suggestion
        maxiter = 10 * tree_size(b)  # copied from scipy
```

File: posteriors/utils.py [hunk 5]
```diff
@@ -228,6 +230,137 @@ def fisher(*args, **kwargs):
...
+    return sum(tree_leaves(tree_map(_vdot_real_part, x, y)))
+
+
+def _mul(scalar, tree) -> TensorTree:
+    return tree_map(partial(operator.mul, scalar), tree)
+
+
+_add = partial(tree_map, operator.add)
+_sub = partial(tree_map, operator.sub)
+_identity = partial(lambda x: x)
```
Comment (SamDuffield) [inline posteriors/utils.py:266]: is the `partial` doing anything here? if it passes the tests with `def identity(x): return x` then let's use that instead

File: posteriors/utils.py [hunk 6]
```diff
@@ -228,6 +230,137 @@ def fisher(*args, **kwargs):
...
+
+    Returns:
+        The result vector dot-product, a real float
+    """
+    # all our uses of vdot() in CG are for computing an operator of the form
+    #  z^H M z
+    #  where M is positive definite and Hermitian, so the result is
+    # real valued:
+    # https://en.wikipedia.org/wiki/Definiteness_of_a_matrix#Definitions_for_complex_matrices
+    real_part = torch.einsum("...i,...i->...", x.real, y.real)
```
Comment (SamDuffield) [inline posteriors/utils.py:249]: why `einsum` and not `torch.vdot`?

File: posteriors/utils.py [hunk 7]
```diff
@@ -228,6 +230,137 @@ def fisher(*args, **kwargs):
...
+    atol: Tensor = torch.Tensor([0]),
+    M: Callable = _identity,
+) -> Tuple[TensorTree, Any]:
+    """Use Conjugate Gradient iteration to solve ``Ax = b``.
+    ``A`` is suplied as a function instead of a sparse
+    matrix or ``LinearOperator``.
+
+    Derivatives of ``cg`` are implemented via implicit differentiation with
+    another ``cg`` solve, rather than by differentiating *through* the solver.
+    They will be accurate only if both solves converge.
```
Comment (SamDuffield) [inline posteriors/utils.py:286]: Is this comment on implicit differentiation true for our implementation? Maybe just remove it

File: posteriors/utils.py [hunk 8]
```diff
@@ -228,6 +230,137 @@ def fisher(*args, **kwargs):
...
+    *,
+    maxiter: int = None,
+    damping: float = 0,
+    tol: Tensor = torch.Tensor([1e-5]),
+    atol: Tensor = torch.Tensor([0]),
+    M: Callable = _identity,
+) -> Tuple[TensorTree, Any]:
+    """Use Conjugate Gradient iteration to solve ``Ax = b``.
+    ``A`` is suplied as a function instead of a sparse
+    matrix or ``LinearOperator``.
```
Comment (SamDuffield) [inline posteriors/utils.py:282]: ``LinearOperator`` not defined for us

File: posteriors/utils.py [hunk 9]
```diff
@@ -228,6 +230,137 @@ def fisher(*args, **kwargs):
...
+    x0: TensorTree = None,
+    *,
+    maxiter: int = None,
+    damping: float = 0,
+    tol: Tensor = torch.Tensor([1e-5]),
+    atol: Tensor = torch.Tensor([0]),
+    M: Callable = _identity,
+) -> Tuple[TensorTree, Any]:
+    """Use Conjugate Gradient iteration to solve ``Ax = b``.
+    ``A`` is suplied as a function instead of a sparse
```
Comment (SamDuffield) [inline posteriors/utils.py:281]:
```suggestion
    ``A`` is supplied as a function instead of a sparse
```

File: posteriors/utils.py (hunk 6 above)
Comment (KaelanDt) [reply posteriors/utils.py:249]: torch vdot doesn't work with arrays that are more than 1D (see note here: https://pytorch.org/docs/stable/generated/torch.vdot.html)

File: posteriors/utils.py (hunk 2 above)
Comment (KaelanDt) [reply posteriors/utils.py:276]: has to be a tensor for operations later on. I don't like it either will transform it inside the function instead

**Comment by SamDuffield** [comment]
Linking to CG discussion #67

File: posteriors/utils.py [hunk 10]
```diff
@@ -228,6 +230,136 @@ def fisher(*args, **kwargs):
...
+        x0: Starting guess for the solution. Must have the same structure as ``b``.
+        maxiter : Maximum number of iterations.  Iteration will stop after maxiter
+            steps even if the specified tolerance has not been achieved.
+        damping : damping term for the mvp function. Acts as regularization.
+        tol: Tolerance for convergence.
+        atol: Tolerance for convergence. ``norm(residual) <= max(tol*norm(b), atol)``.
+            The behaviour will differ from SciPy unless you explicitly pass
+            ``atol`` to SciPy's ``cg``.
+        M : Preconditioner for A.
+            See [the preconditioned CG method.](https://en.wikipedia.org/wiki/Conjugate_gradient_method#The_preconditioned_conjugate_gradient_method)
```
Comment (SamDuffield) [inline posteriors/utils.py:303]:
```suggestion
        maxiter: Maximum number of iterations.  Iteration will stop after maxiter
            steps even if the specified tolerance has not been achieved.
        damping: damping term for the mvp function. Acts as regularization.
        tol: Tolerance for convergence.
        atol: Tolerance for convergence. ``norm(residual) <= max(tol*norm(b), atol)``.
            The behaviour will differ from SciPy unless you explicitly pass
            ``atol`` to SciPy's ``cg``.
        M: Preconditioner for A.
            See [the preconditioned CG method.](https://en.wikipedia.org/wiki/Conjugate_gradient_method#The_preconditioned_conjugate_gradient_method)
```

**Review by SamDuffield — APPROVED** [review]
